pkg/jobs/privilege/privilege.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/roleoption"
+	"github.com/cockroachdb/errors"
+)
+
+// An AccessLevel is used to indicate how strict an authorization check should
+// be.
+type AccessLevel int
+
+const (
+	// ViewAccess is used to perform authorization for viewing jobs (ex. SHOW JOBS).
+	ViewAccess AccessLevel = iota
+
+	// ControlAccess is used to perform authorization for modifying jobs (ex. PAUSE|CANCEL|RESUME JOB).
+	// This access level performs stricter checks than ViewAccess.
+	//
+	// The set of jobs visible via ControlAccess is a subset of jobs visible via
+	// ViewAccess. In other words: if a user with a given set of privileges is
+	// authorized to modify a job using ControlAccess, they will be authorized to
+	// view it using ViewAccess.
+	ControlAccess
+)
+
+// AuthorizationAccessor is an interface for checking authorization on jobs.
+type AuthorizationAccessor interface {
+	// CheckPrivilegeForTableID mirrors sql.AuthorizationAccessor.
+	CheckPrivilegeForTableID(ctx context.Context, tableID descpb.ID, privilege privilege.Kind) error
+
+	// HasRoleOption mirrors sql.AuthorizationAccessor.
+	HasRoleOption(ctx context.Context, roleOption roleoption.Option) (bool, error)
+
+	// UserHasAdminRole mirrors sql.AuthorizationAccessor.
+	UserHasAdminRole(ctx context.Context, user username.SQLUsername) (bool, error)
+
+	// HasAdminRole mirrors sql.AuthorizationAccessor.
+	HasAdminRole(ctx context.Context) (bool, error)
+
+	// User mirrors sql.PlanHookState.
+	User() username.SQLUsername
+}
+
+// changefeedPrivilegeCheck determines if a user has access to the changefeed defined
+// by the supplied payload.
+func changefeedPrivilegeCheck(
+	ctx context.Context, a AuthorizationAccessor, specs []jobspb.ChangefeedTargetSpecification,
+) error {
+
+	for _, spec := range specs {
+		err := a.CheckPrivilegeForTableID(ctx, spec.TableID, privilege.CHANGEFEED)
+		if err != nil {
+			// When performing SHOW JOBS or SHOW CHANGEFEED JOBS, there may be old changefeed
+			// records that reference tables which have been dropped or are being
+			// dropped. In this case, we would prefer to skip the permissions check on
+			// the dropped descriptor.
+			if pgerror.GetPGCode(err) == pgcode.UndefinedTable || errors.Is(err, catalog.ErrDescriptorDropped) {
+				continue
+			}
+
+			return err
+		}
+	}
+	return nil
+}
+
+// Authorize returns nil if the user is authorized to access the job.
+// If the user is not authorized, then a pgcode.InsufficientPrivilege
+// error will be returned.
+func Authorize(
+	ctx context.Context,
+	a AuthorizationAccessor,
+	jobID jobspb.JobID,
+	payload *jobspb.Payload,
+	accessLevel AccessLevel,
+) error {
+	userIsAdmin, err := a.HasAdminRole(ctx)
+	if err != nil {
+		return err
+	}
```

Comment:
why do we even need to continue this check if we you're and admin? shouldn't we just early return?

Reply:
Makes sense. Just pushed an update.